crates/wasi-common/src/sandboxed_tty_writer.rs

```diff
@@ -0,0 +1,178 @@
+use std::io::{Result, Write};
+
+/// An adapter around a `Write` stream that guarantees that its output
+/// is valid UTF-8 and contains no control characters. It does this by
+/// replacing characters with inert control pictures and replacement
+/// characters.
+pub(crate) struct SandboxedTTYWriter<'writer, Writer>
+where
+    Writer: Write,
+{
+    inner: &'writer mut Writer,
+}
+
+impl<'writer, Writer> SandboxedTTYWriter<'writer, Writer>
+where
+    Writer: Write,
+{
+    /// Construct a new `SandboxedTTYWriter` with the given inner `Writer`.
+    pub(crate) fn new(inner: &'writer mut Writer) -> Self {
+        Self { inner }
+    }
+
+    /// Write a single character to the output.
+    pub(crate) fn write_char(&mut self, c: char) -> Result<()> {
+        self.inner.write(
+            match c {
+                '\u{0000}' => '␀',
+                '\u{0001}' => '␁',
+                '\u{0002}' => '␂',
+                '\u{0003}' => '␃',
+                '\u{0004}' => '␄',
+                '\u{0005}' => '␅',
+                '\u{0006}' => '␆',
+                '\u{0007}' => '␇',
+                '\u{0008}' => '␈',
+                '\u{0009}' => '\t',
+                '\u{000A}' => '\n',
+                '\u{000B}' => '␋',
+                '\u{000C}' => '␌',
+                '\u{000D}' => '\r',
+                '\u{000E}' => '␎',
+                '\u{000F}' => '␏',
+                '\u{0010}' => '␐',
+                '\u{0011}' => '␑',
+                '\u{0012}' => '␒',
+                '\u{0013}' => '␓',
+                '\u{0014}' => '␔',
+                '\u{0015}' => '␕',
+                '\u{0016}' => '␖',
+                '\u{0017}' => '␗',
+                '\u{0018}' => '␘',
+                '\u{0019}' => '␙',
+                '\u{001A}' => '␚',
+                '\u{001B}' => '␛',
+                '\u{001C}' => '␜',
+                '\u{001D}' => '␝',
+                '\u{001E}' => '␞',
+                '\u{001F}' => '␟',
+                '\u{007F}' => '␡',
+                x if x.is_control() => '�',
+                x => x,
+            }
+            .encode_utf8(&mut [0; 4]) // UTF-8 encoding of a `char` is at most 4 bytes.
```

Comment:
FYI Limit is only **4** if character code follows Unicode std limit 0x10ffff. But ISO std defines limit as 0x7ffffff. Which would require 6 bytes. So depends which limit the implementation ensures. But for simple pass-through encoding 6 bytes would be the safest.

Reply:
Rust [documents](https://doc.rust-lang.org/std/primitive.char.html#panics-2) that 4 is enough. If 5-byte or 6-byte UTF-8 encodings should ever be revived, we can expect Rust to add new functions rather than change such an invariant in existing functions.